streamlink/streamlink
# CI/CD: Github actions

## Why

We're currently using two CI providers, Appveyor for running tests on Windows and TravisCI for everything else. While Appveyor has caused issues in the past and is usually much slower than TravisCI, TravisCI's Windows VMs are still in beta with lack of support for secret env vars. See my previous PR #2342.

Github now supports CI/CD workflows via Github actions on Linux, Windows and macOS, which allows us to have all CI tasks being run in one place with direct integration into Github.

Docs can be found here:
https://help.github.com/en/categories/automating-your-workflow-with-github-actions

## Checklist

- [ ] **Re-enable actions**
  I've disabled all actions in the repo settings for now. These changes should be tested in a secondary repo first. See below...
- [x] **`test` job**
  - [x] Run in parallel on ubuntu-latest, macOS-latest and windows-latest, using Python 2.7 and 3.5..3.8
  - [x] Test (pytest)
  - [x] Lint (flake8)
  - [x] Report code coverage (codecov)
    - [x] Add `CODECOV_TOKEN` to secret env vars
- [x] **`documentation` job** (runs in parallel)
- [x] **`windows-installer` job** (runs in parallel)
  - [x] Deploys build artifact on tag with current version
  - [x] Deploys build artifact on scheduled run with current version
- [x] **`deploy-documentation` job**
  Depends on `test`, `documentation` and `windows-installer`.
  Runs when pushed to master or a tag gets pushed. Forks are disabled.
  - [x] Rewrite as bash script with similar/same behavior as doctr
  - [x] Add secret env vars to repo settings
    - [x] `DOCS_DEPLOY_PASSPHRASE`
  - [x] Replace deploy key
    - [x] Add public key to @streamlink-bot
  - [x] Give @streamlink-bot write access to docs repo
  - [x] Add 2FA to @streamlink-bot 
- [x] **`release` job**
  Depends on `deploy-documentation`.
  Runs if tag was pushed. Forks are disabled.
  - [x] Add secret env vars to repo settings
    - [x] `RELEASES_API_KEY`
    - [x] `RELEASE_SIGN_PASSPHRASE`
    - [x] `PYPI_USER`
    - [x] `PYPI_PASSWORD`
  - [x] Replace signing key (same key, different encryption)
  - [x] Deploy to PyPI
    - [x] Split sdist/wheel script into build and deploy script
  - [x] Deploy to Github releases
    - [x] Cache Windows installer artifact
    - [x] Rewrite Github releases helper script with the following features
      - Create or update release
      - Upload assets
      - Read Github's env vars
      - Move release template text into its own file
- [x] **Verify all this in a secondary repo**
  See https://github.com/bastimeyer-test/streamlink-test/actions?query=is%3Asuccess
  - [x] Test job
    Codecov: https://codecov.io/gh/bastimeyer-test/streamlink-test/commit/c10c200520548a27e7fbf08c13709cea23b3b7e7
  - [x] Documentation job
  - [x] Windows installer job
  - [x] Deploy documentation job
    - Latest
      - https://github.com/bastimeyer-test/streamlink-test/runs/557448378
      - https://github.com/bastimeyer-test/streamlink-website-test/commit/d294d345fd6d6c5a6535f82572ef90b49af8f64d
    - Stable
      - https://github.com/bastimeyer-test/streamlink-test/runs/557465088
      - https://github.com/bastimeyer-test/streamlink-website-test/commit/2cdb908a58351d4835288781b1640550191db22c
  - [x] Release job
    - https://github.com/bastimeyer-test/streamlink-test/runs/557468060
    - https://github.com/bastimeyer-test/streamlink-test/releases/tag/2020.4.3
  - [x] Scheduled nightly build
    - https://github.com/bastimeyer-test/streamlink-test/runs/557471458
    - https://github.com/bastimeyer-test/streamlink-test/actions/runs/69723359
- [ ] **Remove old stuff once approved**
  - [ ] TravisCI config
  - [ ] AppVeyor config
  - [x] Old doctr deploy keys
  - [ ] Unused bintray deploy scripts

## Questions

- **How fast is it?**
  Build time changed from >15 mins to <5 mins.
- **Is testing on macOS really needed?**
  This spawns 5 additional jobs, but doesn't slow down the build, as the jobs running on Windows always finish last. Adds to the repo's overall available build time though.
- **Build artifacts**
  The windows installer is now stored as a build artifact, which will be deleted after 90 days. There's an API endpoint available that can be queried for getting the latest artifact. This could be added to the docs, or just a link to the repo's successfully run scheduled actions.
